### DISCUSSION TITLE
Installing OpenCV with CUDA for Python

### DISCUSSION THREAD

**Discussion by Petopp** [ifm/ifm3d#383] [Q&A] [answered]
Hello all,
I'm just getting to grips with O3R and thanks to some help I've been able to get a lot of things working.

The only thing I'm really struggling with is CUDA and OpenCV. I get an OpenCV created with Cuda, but when copying (multistage) the created data I blow up the maximum of the docker which works on the devices. 

So the question is, has anyone already done this, and if so, how?

Here's what I've got so far, but it takes a while to execute and it's more work in process at the moment:

```
FROM ghcr.io/ifm/ifm3d:latest-l4t-arm64 AS buildstage

COPY requirements.txt /tmp/
COPY python /tmp/python

USER root

ARG JETPACK_VERSION_BASE="r32.4"
ARG JETPACK_VERSION="${JETPACK_VERSION_BASE}.3"
ARG BASE_IMAGE="nvcr.io/nvidia/l4t-base:${JETPACK_VERSION}"
ARG SOC="t186"

ADD --chown=root:root https://repo.download.nvidia.com/jetson/jetson-ota-public.asc /etc/apt/trusted.gpg.d/jetson-ota-public.asc
RUN chmod 644 /etc/apt/trusted.gpg.d/jetson-ota-public.asc \
    && apt-get update && apt-get install -y --no-install-recommends \
        ca-certificates \
    && echo "deb https://repo.download.nvidia.com/jetson/common ${JETPACK_VERSION_BASE} main" > /etc/apt/sources.list.d/nvidia-l4t-apt-source.list \
    && echo "deb https://repo.download.nvidia.com/jetson/${SOC} ${JETPACK_VERSION_BASE} main" >> /etc/apt/sources.list.d/nvidia-l4t-apt-source.list \
    && cat /etc/apt/sources.list.d/nvidia-l4t-apt-source.list \
    && apt-get update \
    && rm -rf /var/lib/apt/lists/*


ARG CUDA=10.2
ENV CUDA=${CUDA}

ENV PATH /usr/local/cuda-$CUDA/bin:/usr/local/cuda/bin:${PATH}
ENV LD_LIBRARY_PATH /usr/local/cuda-$CUDA/targets/aarch64-linux/lib:${LD_LIBRARY_PATH}

RUN ldconfig

ENV NVIDIA_VISIBLE_DEVICES all
ENV NVIDIA_DRIVER_CAPABILITIES all


ARG OPENCV_VERSION=4.7.0


RUN apt-get update && \
    apt-get install -y \
    bc \
    bzip2 \
    language-pack-en-base \
    python3-distutils \
    python3-pip \
    build-essential \
    cuda-libraries-dev-${CUDA} \
    cuda-cudart-dev-${CUDA} \
    cuda-compiler-${CUDA} \
    libnvinfer-samples \
    ca-certificates \
    python-dev \
    git \
    cmake \
    wget \
        unzip \
        yasm \
        pkg-config \
        libswscale-dev \
        libtbb2 \
        libtbb-dev \
        libjpeg-dev \
        libpng-dev \
        libtiff-dev \
        libavformat-dev \
        libpq-dev \
        libxine2-dev \
        libglew-dev \
        libtiff5-dev \
        zlib1g-dev \
        libjpeg-dev \
        libavcodec-dev \
        libavformat-dev \
        libavutil-dev \
        libpostproc-dev \
        libswscale-dev \
        libeigen3-dev \
        libtbb-dev \
        libgtk2.0-dev \
        pkg-config \
        ## Python
        python3-numpy \
    && rm -rf /var/lib/apt/lists/*


USER ifm
RUN python3 -m pip install --upgrade pip
RUN pip3 install --requirement /tmp/requirements.txt

USER root

RUN cd /opt/ &&\
    # Download and unzip OpenCV and opencv_contrib and delte zip files
    wget https://github.com/opencv/opencv/archive/$OPENCV_VERSION.zip &&\
    unzip $OPENCV_VERSION.zip &&\
    rm $OPENCV_VERSION.zip &&\
    wget https://github.com/opencv/opencv_contrib/archive/$OPENCV_VERSION.zip &&\
    unzip ${OPENCV_VERSION}.zip &&\
    rm ${OPENCV_VERSION}.zip &&\
    # Create build folder and switch to it
    mkdir /opt/opencv-${OPENCV_VERSION}/build && cd /opt/opencv-${OPENCV_VERSION}/build &&\
    # Cmake configure
    cmake \
        -DOPENCV_EXTRA_MODULES_PATH=/opt/opencv_contrib-${OPENCV_VERSION}/modules \
        -DWITH_CUDA=ON \
		-DWITH_BUILD_JAVA=OFF \
		-DWITH_BUILD_opencv_python2=OFF \
		-DWITH_BUILD_SHARED_LIBS=OFF \
        -DCMAKE_BUILD_TYPE=RELEASE \
        # Install path will be /usr/local/lib (lib is implicit)
        -DCMAKE_INSTALL_PREFIX=/usr/local \
        .. &&\
    # Make
    make -j"$(nproc)" && \
    # Install to /usr/local/lib
    make install && \
    ldconfig &&\
    # Remove OpenCV sources and build folder
    rm -rf /opt/opencv-${OPENCV_VERSION} && rm -rf /opt/opencv_contrib-${OPENCV_VERSION}


#CMD ["/bin/bash"]


RUN ls /usr/local/lib/libopencv*

FROM ghcr.io/ifm/ifm3d:latest-l4t-arm64

USER root


#COPY --from=buildstage /usr/local/include/opencv4/ /usr/local/include/opencv4/
COPY --from=buildstage /home/ifm/venv/lib/python3.9/site-packages /home/ifm/venv/lib/python3.9/site-packages
COPY --from=buildstage /usr/local/lib/python3.9/site-packages/cv2/ /home/ifm/venv/lib/python3.9/site-packages/cv2/
COPY --from=buildstage /usr/local/lib/libopencv* /usr/local/lib


COPY --from=buildstage /usr/lib/x86_64-linux-gnu/ /usr/lib/x86_64-linux-gnu/


#RUN ldconfig
#COPY python /tmp/python
```

**Comment by Stedag** [comment]
Hi, do you know the size of the resulting image if exported as a tarball?

**Comment by Petopp** [comment]
Hello,
the file tar file size it's 1,31GB.

**Comment by Stedag** [comment]
This size should fit on the VPU.

Can you try:
1. Removing any existing .tar files from the VPU
2. Using `docker images purge` to free up space

**Comment by Stedag** [comment]
Additionally, my understanding is that if you pipe the image directly into docker that the image remains in memory. This may work around storage limitations...

`docker save yourimage | gzip | ssh -C oem@IP_ADDR docker load`

**Comment by Petopp** [comment]
Hello,
I use that deletion as well as the way of transmission. Unfortunately, an error message still appears.

I have now also reset everything on the build computer and am now creating a new container.

I will then try to transfer it. 
What is the maximum size for a container?

**Reply by Stedag** [reply]
I don't have an exact number available offhand. I'm going to run some tests and try to get back to you.

**Comment by Stedag** [comment]
There should be about 3GB available on a newly flashed VPU, This should allow containers up to ~2.5 GB. Can you paste the results of typing `df -h` via ssh on the VPU?

**Comment by Petopp** [comment]
Hello,
I have redone the container, but unfortunately the problem remains.
Here is the exact Error Message:
![image](https://user-images.githubusercontent.com/41972011/216146569-4956ae96-d98c-46b9-a4ff-31f3a710db1d.png)


Here the desired overview
![image](https://user-images.githubusercontent.com/41972011/216145961-dfdcbf35-fbfd-4159-a52e-849ecf982cef.png)

**Comment by Stedag** [comment]
This has actually happened to me before. The internal development team is working on this issue. Currently the recourse is to reflash the firmware of the VPU.

If you have not done so previously, It may be a good idea to update the firmware on your VPU anyway. Here are the instructions: https://ifm3d.com/documentation/Firmware/fw_update.html

**Reply by Petopp** [reply]
Hello, i have flasched the firmware. The memory it's now:
![image](https://user-images.githubusercontent.com/41972011/216261592-f82eb7be-9409-4cea-a9d1-4bafd8666b9a.png)

However, the problem is still there by copy the container to the device.

**Comment by Petopp** [comment]
Hello,
Another solution for me would be to first install this framework https://github.com/pjreddie/darknet for YOLO with CUDA support on the device, if there is enough memory. Have any one experience with this?

**Comment by graugans** [comment, marked as answer]
Hej @Petopp as @Stedag mentioned. We are working on a firmware update which provides more space for containers. It may makes sense to reach out to support.robotics@ifm.com to discuss your project and timelines in detail.

**Reply by Petopp** [reply]
Hello, I will contact you. Maybe we will find another solution. thanks for the help

**Reply by graugans** [reply]
Hey @Petopp I tried to build your container for testing and stumbled across this line:

> COPY --from=buildstage /usr/lib/x86_64-linux-gnu/ /usr/lib/x86_64-linux-gnu/
> failed to compute cache key: "/usr/lib/x86_64-linux-gnu" not found: not found

Which makes totally sense due to the fact we are building for `arm64`

**Reply by Petopp** [reply]
Hello, you are right, but when testing, opencv asked for the data from the directory. It may also be that this is not necessary if you provide the lib folder and its data. Unfortunately I could not test this.

**Reply by graugans** [reply]
I also figured out that my container is 2.72 GB in size:

````
docker build -t ifm3d/aarch64/opencv4:latest . 
docker image ls | grep opencv4
ifm3d/aarch64/opencv4                            latest                656d92caff15   4 minutes ago   2.72GB
````

**Reply by graugans** [reply]
````
docker save ifm3d/aarch64/opencv4 > test.tar &&  du -h test.tar && rm test.tar
2,6G    test.tar
````

**Reply by Petopp** [reply]
Yes, i have the same size today. I'm also a little confused right now. I may have used the wrong build where the lib directory was missing.

**Reply by graugans** [reply]
Those numbers make more sense cause the base image you are using already consumes 1.02GB

```
ghcr.io/ifm/ifm3d                                latest-l4t-arm64      6c69bb1884c6   N/A              1.02GB
````
Maybe the image can be further reduced by optimizing the base image and removing everything which is not really needed. This is will consume some amount of compile time I guess....

**Reply by Petopp** [reply]
I'm afraid there's just not much memory left to save. Since CUDA as such already needs quite a bit of memory. I just had a look at a few other containers in the Docker Hub, and in terms of memory, anything with CUDA under 1GB is unlikely to be found.

**Reply by graugans** [reply]
Hm, CUDA shouldn't be needed in the final Docker container. This is mounted from the O3R host. It is only a build time dependency. I guess the ifm3d Docker container has gcc installed, this can also be a possibility to safe some memory footprint. @inbangsa mentioned that you might can also reduce the OpenCV modules to a smaller subgroup.

**Reply by Petopp** [reply]
Hi, the problem with the reduce solution is that a lot of files are needed by openCV to run with CUDA. These are the files which are located under /usr/local/lib/libopencv*. There may be even more. I have not yet been able to find more details on this.

The only solution I have right now is to run it on CPU. But that doesn't help me with this project, because it's much too slow. 

I think we will wait for the new firmware and then see if the memory has improved.

Alternatively, we could use pytorch, but that's another chapter.

**Reply by graugans** [reply]
May I am mixing things up a little bit but Pytorch is for machine learning as far as I know the OpenCV stuff is more like classical image recognition. In terms of AI/ML there is the TensorRT library from Nvidia. You also have to take care that the 3D image data is relatively small and moving data to and from the GPU may also consume time. So offloading a single filter in CUDA may not bring the performance gain you want to see.

**Reply by Petopp** [reply]
yes, it is easy to mix it up too much. With Opencv it is possible to load and process ML models such as YOLO or ONNX files. When processing, you can also load CUDA support. This increases the throughput considerably, e.g. even an old GT 720 can sometimes be faster than a PC when processing models and consume less power. So CUDA helps a lot here!

It's also only about 2D image recognition in my project, 3D data is sufficient here as values (fortunately :-) ).

Pytorch as well as Tensorflow have the possibility to run the models directly. Installing CUDA is easier in both cases, but they also need more memory than the board currently has. 

Tensorflow RT would be an alternative, which I will take a look at next week. I'm on a business trip this week and can't continue there at the moment.

**Reply by graugans** [reply]
Okay, thanks for the details.
